README.md

```diff
@@ -14,8 +10,279 @@ But we need have a clear way to discover how you've been invited to a group with
 We also need to consider how to ensure our group data is replicated _enough_ - e.g. if a group is just me and my sister, then are there gonna be enough copies of the group data to be resilent/ timely.
 
 
---- 
+## Principles
+
+1. **group membership should be opaque**
+   - you shouldn't be able to guess who is in a group using public info
+     - shards people are using
+     - groupId (e.g. %a2148bs2K3klmhnaALsdaN90=.cloaked)
+2. **we support sympathetic replication**
+
+
+
+## Spec
+
+This work builds on the [ssb-meta-feeds-spec] (v1).
+
+We define two types of feeds that each peer will have:
+1. <div class='group-invite'></div> the group-invite feed
+2. <div class='group'></div>        group feeds
+
+
+```mermaid
+graph TB
+
+root(root)
+v1(v1)
+5(5) & b(b) & f(f)
+group-invite(group-invite):::groupInvite
+denmark(denmark):::group
+helsinki(helsinki):::group
+wellington(wellington):::group
+
+root --> v1 --> 5 --> helsinki
+         v1 --> b --> group-invite
+         v1 --> f --> denmark
+                f --> wellington
+
+classDef default stroke:none;
+classDef groupInvite fill: #BF2669, stroke:none, color:white;
+classDef group fill: #702A8C, stroke: none, color:white;
+```
+_Diagram showing an example layout of group-related feeds. Note that the shards in your
+use-case will likely not be those shown, see how they are determined below._
+
+
+### the group-invite feed
+
+This feed holds messages which help peers join groups (e.g. `group/registration`, `group/add-member` messages).
+This feed MUST be unique for each peer (a singleton).
+Each peer MUST replicate this feed from each of their peers (OR each peer they might possibly enter a group with).
+
+For each peer, this feed MUST be located on the meta-feed tree at:
+```
+/v1/:shard/group-invite
+```
+Where:
+- `:shard` is the shard feed derived using the string `"group-invite"`
+- `group-invite` is a content feed with
+    - `feedpurpose = "group-invite"`
+    - `feedtype = ?` :warning:
+
+See [ssb-meta-feeds-spec] for detail about the `v1` shared and the `shard` calculation.
+
+All content on this feed SHOULD be encrypted with box2 encryption.
+
+
+### group feeds
+
+Group feeds MUST be located on the meta-feed tree at:
+
+```
+/v1/:shard/:groupKey
+```
+Where:
+- `:shard` is the shard feed derived using BFE encoded `groupKey`
+    <details>
+      <summary>details</summary>
+      <div>
+        We cannot use the `groupId`, as this is publicly known, which would give attackers a way to test if people are in the group (breaking Principle 1.)
```

Comment:
I had some realisations about the "uniqueName" string thing used to determine a shared:
1. it cannot be the groupId - this is public already, and would allow people to sniff out group members (somewhat)
2. it cannot be just any string - unless a particular app just runs in a global singleton group .... which I cannot think of any usecase for, it would essentially be public to anyone with the app...

So it has to be private to people in the group, and it has to be unique... that leaves us with the groupKey?

I'm noticing we're doing "crypto uniqueness" things here, and that Kek's brain would be **invaluable** for checking these assumptions. I can spot _some_  problems but I would consider myself a beginner and we should seek professional advice. (i.e. a security auditor)

Reply:
Yes, agreed. And I'm grateful you were able to spot this groupId versus groupKey dilemma, it's important, and wasn't obvious to me. There are details here that we didn't foresee.

Reply:
What if this was a random 32 bit value created when you start the group. Then all invited would know this and we can use it for sympathy replication by revealing it to those that would like to help triplicate the group without knowing it's content

Reply:
you're proposing adding a new thing to a `group/add-member` message?

Problem - what if you add 4 people at once?
Solution - not a random 32bits, instead  `hash(groupAddMsgId, userId)`